File: plugins/openclaw_agentoracle_plugin/src/websocket_client.py
"""
WebSocket Client for OpenClaw Gateway (Protocol v3)

This module handles WebSocket communication with OpenClaw Gateway using Protocol v3.
It provides a production-ready implementation with automatic reconnection, exponential
backoff, and comprehensive error handling.

Based on openclaw_daily_elf implementation.
"""

import asyncio
import websockets
import json
import uuid
from typing import Optional
from datetime import datetime

# Support both module and direct execution
try:
    from .logger import setup_logger
except ImportError:
    from logger import setup_logger


class OpenClawWebSocketClient:
    """WebSocket client for OpenClaw Gateway Protocol v3"""
    
    def __init__(self, 
                 gateway_url: str = "ws://127.0.0.1:18789",
                 gateway_token: Optional[str] = None,
                 timeout: int = 300,
                 max_retries: int = 3,
                 connect_timeout: int = 10,
                 message_timeout: int = 20):
        """Initialize WebSocket client.
        
        Args:
            gateway_url: WebSocket URL for OpenClaw Gateway (default: ws://127.0.0.1:18789)
            gateway_token: Authentication token for Gateway (optional)
            timeout: Total timeout for message exchange in seconds (default: 300)
            max_retries: Maximum number of retry attempts (default: 3)
            connect_timeout: Connection timeout in seconds (default: 10)
            message_timeout: Single message timeout in seconds (default: 20)
        """
        self.gateway_url = gateway_url
        self.gateway_token = gateway_token
        self.timeout = timeout
        self.max_retries = max_retries
        self.connect_timeout = connect_timeout
        self.message_timeout = message_timeout
        self.logger = setup_logger()
        
        # Retry delay base (exponential backoff)
        self.retry_delay_base = 2
    
    async def send_message(self, message: str) -> Optional[str]:
        """Send message to OpenClaw Gateway and receive response.
        
        Implements Protocol v3 handshake:
        1. Receive connect.challenge
        2. Send connect request with authentication
        3. Receive connect response
        4. Send chat.send message
        5. Receive chat events until completion
        
        Includes automatic reconnection with exponential backoff.
        
        Args:
            message: Message text to send to the agent
            
        Returns:
            Agent's response text, or None if communication failed
        """
        for attempt in range(self.max_retries):
            try:
                if attempt > 0:
                    # Exponential backoff
                    delay = self.retry_delay_base ** attempt
                    self.logger.info(f"[AgentOracle] 🔄 第 {attempt + 1}/{self.max_retries} 次重试，等待 {delay} 秒...")
                    await asyncio.sleep(delay)
                
                self.logger.info(f"[AgentOracle] 🔌 连接到 {self.gateway_url}...")
                
                async with websockets.connect(
                    self.gateway_url,
                    ping_interval=30,  # 每30秒发送心跳
                    ping_timeout=20,   # 20秒无响应则超时
                    close_timeout=10   # 关闭超时
                ) as websocket:
                    self.logger.info("[AgentOracle] ✅ 已连接到 OpenClaw Gateway")
                    
                    # Step 1: Receive connect.challenge
                    self.logger.info("[AgentOracle] ⏳ 等待 connect.challenge...")
                    challenge_msg = await asyncio.wait_for(
                        websocket.recv(), 
                        timeout=self.connect_timeout
                    )
                    challenge_data = json.loads(challenge_msg)
                    self.logger.info(f"[AgentOracle] 📨 收到 challenge")
                    
                    if challenge_data.get("type") != "event" or challenge_data.get("event") != "connect.challenge":
                        raise Exception(f"意外的消息类型: {challenge_data.get('type')}")
                    
                    nonce = challenge_data.get("payload", {}).get("nonce")
                    if not nonce:
                        raise Exception("Challenge 中没有 nonce")
                    
                    # Step 2: Send connect request
                    connect_id = str(uuid.uuid4())
                    connect_request = {
                        "type": "req",
                        "id": connect_id,
                        "method": "connect",
                        "params": {
                            "minProtocol": 3,
                            "maxProtocol": 3,
                            "client": {
                                "id": "cli",  # Must be 'cli' as required by OpenClaw Gateway
                                "version": "1.0.0",
                                "platform": "windows",
                                "mode": "cli"  # Must be 'cli' as required by OpenClaw Gateway
                            },
                            "role": "operator",
                            "scopes": ["operator.read", "operator.write"],
                            "caps": [],
                            "commands": [],
                            "permissions": {},
                            "locale": "zh-CN",
                            "userAgent": "openclaw-agentoracle-plugin/1.0.0"
                        }
                    }
                    
                    # Add authentication if token is provided
                    if self.gateway_token:
                        connect_request["params"]["auth"] = {
                            "token": self.gateway_token
                        }
                    
                    self.logger.info("[AgentOracle] 📤 发送 connect 请求...")
                    await websocket.send(json.dumps(connect_request))
                    
                    # Step 3: Receive connect response
                    self.logger.info("[AgentOracle] ⏳ 等待 connect 响应...")
                    connect_response_msg = await asyncio.wait_for(
                        websocket.recv(), 
                        timeout=self.connect_timeout
                    )
                    connect_response = json.loads(connect_response_msg)
                    
                    if connect_response.get("type") != "res" or not connect_response.get("ok"):
                        error = connect_response.get("error", "未知错误")
                        raise Exception(f"连接失败: {error}")
                    
                    self.logger.info("[AgentOracle] ✅ 连接成功！")
                    
                    # Step 4: Use main session (same as daily_elf)
                    # Extract mainSessionKey from connect response, fallback to agent:main:main
                    session_key = connect_response.get("payload", {}).get("snapshot", {}).get("sessionDefaults", {}).get("mainSessionKey", "agent:main:main")
                    
                    # Log session key for debugging
                    self.logger.info(f"[AgentOracle] 📋 使用 Session Key: {session_key}")
                    
                    chat_id = str(uuid.uuid4())
                    chat_request = {
                        "type": "req",
                        "id": chat_id,
                        "method": "chat.send",
                        "params": {
                            "sessionKey": session_key,
                            "message": message,
                            "idempotencyKey": f"agentoracle-{datetime.now().timestamp()}"
                        }
                    }
                    
                    self.logger.info(f"[AgentOracle] 📤 发送聊天消息（长度: {len(message)} 字符）...")
                    await websocket.send(json.dumps(chat_request))
                    
                    # Step 5: Receive response
                    self.logger.info("[AgentOracle] ⏳ 等待 AI 响应...")
                    full_reply = ""
                    start_time = asyncio.get_event_loop().time()
                    chat_completed = False
                    last_update_time = start_time
                    
                    while True:
                        elapsed = asyncio.get_event_loop().time() - start_time
                        if elapsed > self.timeout:
                            self.logger.error(f"[AgentOracle] ⏰ 总超时 ({self.timeout}秒)")
                            raise Exception(f"任务超时 ({self.timeout}秒)")
                        
                        try:
                            response_msg = await asyncio.wait_for(
                                websocket.recv(), 
                                timeout=self.message_timeout
                            )
                            response = json.loads(response_msg)
                            msg_type = response.get('type', 'unknown')
                            
                            if msg_type == "res":
                                # Response message
                                if response.get("id") == chat_id:
                                    if response.get("ok"):
                                        self.logger.info("[AgentOracle] ✅ 聊天请求已接受")
                                    else:
                                        error = response.get("error", "未知错误")
                                        raise Exception(f"聊天错误: {error}")
                            
                            elif msg_type == "event":
                                # Event message
                                event = response.get("event", "unknown")
                                
                                if event == "chat":
                                    payload = response.get("payload", {})
                                    state = payload.get("state", "")
                                    
                                    # Extract message content
                                    message_data = payload.get("message", {})
                                    if isinstance(message_data, dict):
                                        content_list = message_data.get("content", [])
                                        if content_list and len(content_list) > 0:
                                            text_content = content_list[0].get("text", "")
                                            if text_content:
                                                full_reply = text_content
                                                last_update_time = asyncio.get_event_loop().time()
                                                
                                                # Show progress
                                                reply_len = len(full_reply)
                                                if reply_len > 0:
                                                    self.logger.info(f"[AgentOracle] ✨ 收到更新 (已接收 {reply_len} 字符, state={state})")
                                    
                                    # Check if completed
                                    if state in ["final", "done", "complete", "finished"]:
                                        self.logger.info(f"[AgentOracle] ✅ 聊天完成 (state: {state})")
                                        chat_completed = True
                                        break
                        
                        except asyncio.TimeoutError:
                            # Check if no updates for a long time
                            time_since_update = asyncio.get_event_loop().time() - last_update_time
                            if time_since_update > self.message_timeout * 2:
                                if chat_completed or full_reply:
                                    self.logger.info("[AgentOracle] ✅ 长时间无更新，假定完成")
                                    break
                                else:
                                    self.logger.warning(f"[AgentOracle] ⏰ {self.message_timeout}秒内无消息...")
                                    continue
                            else:
                                continue
                    
                    if full_reply:
                        self.logger.info(f"[AgentOracle] ✅ 成功获取回复 ({len(full_reply)} 字符)")
                        return full_reply
                    else:
                        self.logger.warning("[AgentOracle] ⚠️  未收到回复内容")
                        if attempt < self.max_retries - 1:
                            continue  # Retry
                        return None
                        
            except asyncio.TimeoutError as e:
                self.logger.error(f"[AgentOracle] ⏰ 连接超时: {e}")
                if attempt < self.max_retries - 1:
                    continue  # Retry
                return None
                
            except websockets.exceptions.WebSocketException as e:
                self.logger.error(f"[AgentOracle] 🔌 WebSocket 错误: {e}")
                if attempt < self.max_retries - 1:
                    continue  # Retry
                return None
                
            except Exception as e:
                self.logger.error(f"[AgentOracle] 💥 意外错误: {e}")
                if attempt < self.max_retries - 1:
                    continue  # Retry
                import traceback
                traceback.print_exc()
                return None
        
        self.logger.error(f"[AgentOracle] ❌ 达到最大重试次数 ({self.max_retries})，任务失败")
        return None
    
    def send_message_sync(self, message: str) -> Optional[str]:
        """Synchronous wrapper for send_message.
        
        This method allows calling the async send_message from synchronous code.
        
        Args:
            message: Message text to send
            
        Returns:
            Agent's response text, or None if communication failed
        """
        try:
            # Create new event loop for this call
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            try:
                return loop.run_until_complete(self.send_message(message))
            finally:
                loop.close()
        except Exception as e:
            self.logger.error(f"[AgentOracle] 同步调用失败: {e}", exc_info=True)
            return None
    
    async def test_connection(self) -> bool:
        """Test connection to OpenClaw Gateway.
        
        Returns:
            True if connection successful, False otherwise
        """
        try:
            self.logger.info(f"[AgentOracle] 测试 WebSocket 连接: {self.gateway_url}")
            
            async with websockets.connect(
                self.gateway_url,
                ping_interval=30,
                ping_timeout=20,
                close_timeout=10
            ) as websocket:
                # Wait for challenge
                challenge_msg = await asyncio.wait_for(
                    websocket.recv(),
                    timeout=self.connect_timeout
                )
                challenge_data = json.loads(challenge_msg)
                
                if challenge_data.get("type") == "event" and challenge_data.get("event") == "connect.challenge":
                    self.logger.info("[AgentOracle] ✅ WebSocket 连接测试成功")
                    return True
                else:
                    self.logger.warning("[AgentOracle] ⚠️ 收到意外的消息类型")
                    return False
                    
        except Exception as e:
            self.logger.error(f"[AgentOracle] ❌ WebSocket 连接测试失败: {e}")
            return False
    
    def test_connection_sync(self) -> bool:
        """Synchronous wrapper for test_connection.
        
        Returns:
            True if connection successful, False otherwise
        """
        try:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            try:
                return loop.run_until_complete(self.test_connection())
            finally:
                loop.close()
        except Exception as e:
            self.logger.error(f"[AgentOracle] 连接测试失败: {e}", exc_info=True)
            return False

File: plugins/agentoracle-native-plugin/src/websocket_client.ts
import WebSocket from 'ws';
import { v4 as uuidv4 } from 'uuid';
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import {
  WebSocketConfig,
  ConnectChallengeEvent,
  ConnectRequest,
  ConnectResponse,
  ChatSendRequest,
  ChatEvent,
  DeviceParams,
  PluginLogger,
  NetworkError,
  AuthenticationError
} from './types';

interface DeviceIdentityData {
  deviceId: string;
  publicKey: string;
  privateKey: string;
}

interface LoadedDeviceIdentity {
  deviceId: string;
  publicKeyB64: string;
  keyObject: crypto.KeyObject;
}

function loadOrCreateDeviceIdentity(identityFile: string): LoadedDeviceIdentity | null {
  try {
    let data: DeviceIdentityData;

    if (fs.existsSync(identityFile)) {
      data = JSON.parse(fs.readFileSync(identityFile, 'utf-8')) as DeviceIdentityData;
    } else {
      const { privateKey, publicKey } = crypto.generateKeyPairSync('ed25519');
      const pubJwk = publicKey.export({ format: 'jwk' }) as { x: string };
      const privJwk = privateKey.export({ format: 'jwk' }) as { d: string };

      const pubB64 = pubJwk.x;
      const padding = (4 - (pubB64.length % 4)) % 4;
      const pubBytes = Buffer.from(pubB64 + '='.repeat(padding), 'base64');
      const deviceId = crypto.createHash('sha256').update(pubBytes).digest('hex');

      data = { deviceId, publicKey: pubB64, privateKey: privJwk.d };
      fs.mkdirSync(path.dirname(identityFile), { recursive: true });
      fs.writeFileSync(identityFile, JSON.stringify(data, null, 2), 'utf-8');
    }

    const pubPadding = (4 - (data.publicKey.length % 4)) % 4;
    const pubBytes = Buffer.from(data.publicKey + '='.repeat(pubPadding), 'base64');
    const deviceId = crypto.createHash('sha256').update(pubBytes).digest('hex');

    const keyObject = crypto.createPrivateKey({
      key: { kty: 'OKP', crv: 'Ed25519', x: data.publicKey, d: data.privateKey },
      format: 'jwk',
    });

    return { deviceId, publicKeyB64: data.publicKey, keyObject };
  } catch (err) {
    return null;
  }
}

function buildDeviceParams(
  identity: LoadedDeviceIdentity,
  nonce: string,
  gatewayToken: string,
  platform: string
): DeviceParams {
  const signedAtMs = Date.now();
  const normalizedPlatform = platform.toLowerCase().trim();
  const payload = [
    'v3',
    identity.deviceId,
    'cli',
    'cli',
    'operator',
    'operator.read,operator.write',
    String(signedAtMs),
    gatewayToken,
    nonce,
    normalizedPlatform,
    '',
  ].join('|');

  const signature = crypto
    .sign(null, Buffer.from(payload, 'utf-8'), identity.keyObject)
    .toString('base64url')
    .replace(/=/g, '');

  return {
    id: identity.deviceId,
    publicKey: identity.publicKeyB64,
    signature,
    signedAt: signedAtMs,
    nonce,
  };
}

/**
 * WebSocketClient 类
 * 通过 WebSocket Protocol v3 与 OpenClaw Gateway 通信
 * 
 * 实现功能：
 * - 连接握手和认证
 * - 流式响应累积
 * - 自动重连（指数退避）
 * - 心跳机制
 * - 超时控制
 * - 重连事件通知
 */
export class WebSocketClient {
  private onReconnectCallback?: () => void;
  private lastSuccessfulConnection: number = 0;
  private deviceIdentity: LoadedDeviceIdentity | null = null;

  constructor(
    private config: WebSocketConfig,
    private logger: PluginLogger
  ) {
    const identityFile = config.deviceIdentityFile ??
      path.join(__dirname, '..', 'device_identity.json');
    this.deviceIdentity = loadOrCreateDeviceIdentity(identityFile);
    if (this.deviceIdentity) {
      this.logger.info(`[agentoracle-native-plugin] 🔐 Device Identity 已加载 (ID: ${this.deviceIdentity.deviceId.slice(0, 16)}...)`);
    } else {
      this.logger.warn('[agentoracle-native-plugin] ⚠️ Device Identity 加载失败，operator.write scope 可能被 Gateway 清空');
    }
  }

  /**
   * 设置重连回调
   * 当检测到 Gateway 重启并成功重连后触发
   * @param callback 重连成功后的回调函数
   */
  setOnReconnect(callback: () => void): void {
    this.onReconnectCallback = callback;
  }

  /**
   * 发送消息到 OpenClaw Gateway（不等待回复）
   * 用于发送通知类消息，发送后立即关闭连接，不等待 Agent 回复
   * @param message 要发送的消息内容
   * @returns 成功返回空字符串，失败返回 null
   */
  async sendMessageWithoutResponse(message: string): Promise<string | null> {
    try {
      this.logger.info(`[agentoracle-native-plugin] Connecting to ${this.config.gatewayUrl}...`);

      // 创建 WebSocket 连接
      const ws = await this.connect();

      try {
        // 执行认证流程
        const sessionKey = await this.authenticate(ws);

        // 更新最后成功连接时间
        this.lastSuccessfulConnection = Date.now();

        // 发送聊天消息（不等待回复）
        await this.sendChatMessageWithoutResponse(ws, sessionKey, message);

        this.logger.info(`[agentoracle-native-plugin] Message sent successfully (no response expected)`);

        return ''; // 返回空字符串表示成功
      } finally {
        // 立即关闭连接
        if (ws.readyState === WebSocket.OPEN) {
          ws.close();
        }
      }
    } catch (error) {
      this.logger.error('[agentoracle-native-plugin] WebSocket error', error as Error);
      return null;
    }
  }

  /**
   * 发送消息到 OpenClaw Gateway
   * @param message 要发送的消息内容
   * @returns 完整的响应文本，失败返回 null
   */
  async sendMessage(message: string): Promise<string | null> {
    const isReconnect = this.lastSuccessfulConnection > 0;
    
    for (let attempt = 0; attempt < this.config.maxRetries; attempt++) {
      try {
        if (attempt > 0) {
          // 指数退避
          const delay = Math.pow(this.config.retryDelayBase, attempt);
          this.logger.info(`[agentoracle-native-plugin] WebSocket retry ${attempt + 1}/${this.config.maxRetries}, waiting ${delay}s...`);
          await this.sleep(delay * 1000);
        }

        this.logger.info(`[agentoracle-native-plugin] Connecting to ${this.config.gatewayUrl}...`);

        // 创建 WebSocket 连接
        const ws = await this.connect();

        try {
          // 执行认证流程
          const sessionKey = await this.authenticate(ws);

          // 检测是否为重连（Gateway 重启后的首次成功连接）
          const now = Date.now();
          const timeSinceLastConnection = now - this.lastSuccessfulConnection;
          
          // 如果距离上次成功连接超过 60 秒，且之前有过成功连接，则认为是重连
          if (isReconnect && timeSinceLastConnection > 60000) {
            this.logger.info('[agentoracle-native-plugin] ========================================');
            this.logger.info('[agentoracle-native-plugin] 🔄 Gateway reconnection detected!');
            this.logger.info(`[agentoracle-native-plugin] Time since last connection: ${Math.floor(timeSinceLastConnection / 1000)}s`);
            this.logger.info('[agentoracle-native-plugin] ========================================');
            
            // 触发重连回调
            if (this.onReconnectCallback) {
              this.logger.info('[agentoracle-native-plugin] Triggering reconnect callback...');
              try {
                this.onReconnectCallback();
              } catch (callbackError) {
                this.logger.error('[agentoracle-native-plugin] Error in reconnect callback', callbackError as Error);
              }
            }
          }
          
          // 更新最后成功连接时间
          this.lastSuccessfulConnection = now;

          // 发送聊天消息并接收响应
          const result = await this.sendChatMessage(ws, sessionKey, message);

          this.logger.info(`[agentoracle-native-plugin] Successfully received reply (${result.length} characters)`);

          return result;
        } finally {
          // 确保关闭连接
          if (ws.readyState === WebSocket.OPEN) {
            ws.close();
          }
        }
      } catch (error) {
        this.logger.error('[agentoracle-native-plugin] WebSocket error', error as Error);

        if (attempt < this.config.maxRetries - 1) {
          continue; // 重试
        }

        return null;
      }
    }

    this.logger.error(`[agentoracle-native-plugin] Max retries (${this.config.maxRetries}) reached, task failed`);
    return null;
  }

  /**
   * 测试连接并检测重连
   * 用于在 Gateway 重启后主动检测连接状态
   * @returns 连接是否成功
   */
  async testConnection(): Promise<boolean> {
    const isReconnect = this.lastSuccessfulConnection > 0;
    
    try {
      this.logger.info('[agentoracle-native-plugin] Testing WebSocket connection...');
      
      // 创建 WebSocket 连接
      const ws = await this.connect();

      try {
        // 执行认证流程
        await this.authenticate(ws);

        // 检测是否为重连
        const now = Date.now();
        const timeSinceLastConnection = now - this.lastSuccessfulConnection;
        
        // 如果距离上次成功连接超过 60 秒，且之前有过成功连接，则认为是重连
        if (isReconnect && timeSinceLastConnection > 60000) {
          this.logger.info('[agentoracle-native-plugin] ========================================');
          this.logger.info('[agentoracle-native-plugin] 🔄 Gateway reconnection detected!');
          this.logger.info(`[agentoracle-native-plugin] Time since last connection: ${Math.floor(timeSinceLastConnection / 1000)}s`);
          this.logger.info('[agentoracle-native-plugin] ========================================');
          
          // 触发重连回调
          if (this.onReconnectCallback) {
            this.logger.info('[agentoracle-native-plugin] Triggering reconnect callback...');
            try {
              this.onReconnectCallback();
            } catch (callbackError) {
              this.logger.error('[agentoracle-native-plugin] Error in reconnect callback', callbackError as Error);
            }
          }
        }
        
        // 更新最后成功连接时间
        this.lastSuccessfulConnection = now;

        this.logger.info('[agentoracle-native-plugin] Connection test successful');
        return true;
      } finally {
        // 确保关闭连接
        if (ws.readyState === WebSocket.OPEN) {
          ws.close();
        }
      }
    } catch (error) {
      this.logger.error('[agentoracle-native-plugin] Connection test failed', error as Error);
      return false;
    }
  }

  /**
   * 连接到 Gateway
   * @returns WebSocket 实例
   * @private
   */
  private async connect(): Promise<WebSocket> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(this.config.gatewayUrl, {
        handshakeTimeout: this.config.connectTimeout * 1000
      });

      // 设置心跳
      const pingInterval = setInterval(() => {
        if (ws.readyState === WebSocket.OPEN) {
          ws.ping();
        }
      }, 30000); // 每30秒发送心跳

      ws.on('open', () => {
        this.logger.info('[agentoracle-native-plugin] WebSocket connected');
        resolve(ws);
      });

      ws.on('error', (error) => {
        clearInterval(pingInterval);
        reject(new NetworkError(`WebSocket connection failed: ${error.message}`));
      });

      ws.on('close', () => {
        clearInterval(pingInterval);
      });

      // 连接超时
      setTimeout(() => {
        if (ws.readyState !== WebSocket.OPEN) {
          clearInterval(pingInterval);
          ws.terminate();
          reject(new NetworkError('WebSocket connection timeout'));
        }
      }, this.config.connectTimeout * 1000);
    });
  }

  /**
   * 处理认证流程
   * @param ws WebSocket 实例
   * @returns sessionKey
   * @private
   */
  private async authenticate(ws: WebSocket): Promise<string> {
    // 步骤 1: 接收 connect.challenge
    this.logger.info('[agentoracle-native-plugin] Waiting for connect.challenge...');
    const challengeData = await this.receiveMessage<ConnectChallengeEvent>(
      ws,
      this.config.connectTimeout * 1000
    );

    if (challengeData.type !== 'event' || challengeData.event !== 'connect.challenge') {
      throw new Error(`Unexpected message type: ${challengeData.type}`);
    }

    const nonce = challengeData.payload?.nonce;
    if (!nonce) {
      throw new Error('No nonce in challenge');
    }

    this.logger.info('[agentoracle-native-plugin] Received challenge');

    // 步骤 2: 发送 connect 请求
    const connectId = uuidv4();
    const platform = process.platform.toLowerCase().trim();
    const connectParams: ConnectRequest['params'] = {
      minProtocol: 3,
      maxProtocol: 3,
      client: {
        id: 'cli',
        version: '1.0.0',
        platform,
        mode: 'cli'
      },
      role: 'operator',
      scopes: ['operator.read', 'operator.write'],
      caps: [],
      commands: [],
      permissions: {},
      auth: {
        token: this.config.gatewayToken
      },
      locale: 'zh-CN',
      userAgent: 'agentoracle-native-plugin/1.0.0'
    };

    if (this.deviceIdentity) {
      connectParams.device = buildDeviceParams(
        this.deviceIdentity,
        nonce,
        this.config.gatewayToken,
        platform
      );
    }

    const connectRequest: ConnectRequest = {
      type: 'req',
      id: connectId,
      method: 'connect',
      params: connectParams
    };

    this.logger.info('[agentoracle-native-plugin] Sending connect request...');
    ws.send(JSON.stringify(connectRequest));

    // 步骤 3: 接收 connect 响应
    this.logger.info('[agentoracle-native-plugin] Waiting for connect response...');
    const connectResponse = await this.receiveMessage<ConnectResponse>(
      ws,
      this.config.connectTimeout * 1000
    );

    if (connectResponse.type !== 'res' || !connectResponse.ok) {
      const error = connectResponse.error || 'Unknown error';
      const errorMessage = typeof error === 'object' ? JSON.stringify(error) : String(error);
      if (typeof error === 'object' && (error as any).code === 'NOT_PAIRED') {
        const reqId = (error as any).details?.requestId ?? 'unknown';
        this.logger.error(`[agentoracle-native-plugin] ❌ 设备配对未完成！请在 OpenClaw Control UI 中批准配对请求 ID=${reqId}，然后重启插件`);
        throw new AuthenticationError(`Device pairing required (requestId=${reqId})`);
      }
      this.logger.error(`[agentoracle-native-plugin] Authentication failed: ${errorMessage}`);
      this.logger.error(`[agentoracle-native-plugin] Full response: ${JSON.stringify(connectResponse)}`);
      throw new AuthenticationError(`Connection failed: ${errorMessage}`);
    }

    const sessionKey = connectResponse.payload?.snapshot?.sessionDefaults?.mainSessionKey || 'agent:main:main';

    this.logger.info('[agentoracle-native-plugin] Connection successful');

    return sessionKey;
  }

  /**
   * 发送聊天消息（不等待回复）
   * @param ws WebSocket 实例
   * @param sessionKey 会话密钥
   * @param message 消息内容
   * @private
   */
  private async sendChatMessageWithoutResponse(ws: WebSocket, sessionKey: string, message: string): Promise<void> {
    const chatId = uuidv4();
    const chatRequest: ChatSendRequest = {
      type: 'req',
      id: chatId,
      method: 'chat.send',
      params: {
        sessionKey,
        message,
        idempotencyKey: `agentoracle-${Date.now()}`
      }
    };

    this.logger.info('[agentoracle-native-plugin] Sending chat message (no response expected)...');
    ws.send(JSON.stringify(chatRequest));

    // 等待一小段时间确保消息发送成功，然后立即返回
    await this.sleep(500);
  }

  /**
   * 发送聊天消息
   * @param ws WebSocket 实例
   * @param sessionKey 会话密钥
   * @param message 消息内容
   * @returns 完整的响应文本
   * @private
   */
  private async sendChatMessage(ws: WebSocket, sessionKey: string, message: string): Promise<string> {
    const chatId = uuidv4();
    const chatRequest: ChatSendRequest = {
      type: 'req',
      id: chatId,
      method: 'chat.send',
      params: {
        sessionKey,
        message,
        idempotencyKey: `agentoracle-${Date.now()}`
      }
    };

    this.logger.info('[agentoracle-native-plugin] Sending chat message...');
    ws.send(JSON.stringify(chatRequest));

    // 接收流式响应
    return await this.receiveStreamingResponse(ws, chatId);
  }

  /**
   * 接收流式响应
   * @param ws WebSocket 实例
   * @param chatId 聊天请求ID
   * @returns 完整的响应文本
   * @private
   */
  private async receiveStreamingResponse(ws: WebSocket, chatId: string): Promise<string> {
    this.logger.info('[agentoracle-native-plugin] Waiting for AI response...');

    let fullReply = '';
    const startTime = Date.now();
    let chatCompleted = false;
    let lastUpdateTime = startTime;

    return new Promise((resolve, reject) => {
      const messageHandler = (data: WebSocket.Data) => {
        try {
          const response = JSON.parse(data.toString());
          const msgType = response.type || 'unknown';

          if (msgType === 'res') {
            // 响应消息
            if (response.id === chatId) {
              if (response.ok) {
                this.logger.info('[agentoracle-native-plugin] Chat request accepted');
              } else {
                const error = response.error || 'Unknown error';
                cleanup();
                reject(new Error(`Chat error: ${error}`));
              }
            }
          } else if (msgType === 'event') {
            // 事件消息
            const event = response.event || 'unknown';

            if (event === 'chat') {
              const payload = response.payload || {};
              const state = payload.state || '';

              // 提取消息内容
              const messageData = payload.message || {};
              if (messageData.content && Array.isArray(messageData.content) && messageData.content.length > 0) {
                const textContent = messageData.content[0].text || '';
                if (textContent) {
                  fullReply = textContent;
                  lastUpdateTime = Date.now();

                  // 显示进度
                  if (fullReply.length > 0) {
                    this.logger.info(`[agentoracle-native-plugin] Received update (${fullReply.length} characters, state=${state})`);
                  }
                }
              }

              // 检查是否完成
              if (['final', 'done', 'complete', 'finished'].includes(state)) {
                this.logger.info(`[agentoracle-native-plugin] Chat completed (state: ${state})`);
                chatCompleted = true;
                cleanup();
                resolve(fullReply);
              }
            }
          }
        } catch (error) {
          this.logger.error('[agentoracle-native-plugin] Error parsing message', error as Error);
        }
      };

      const errorHandler = (error: Error) => {
        cleanup();
        reject(new NetworkError(`WebSocket error: ${error.message}`));
      };

      const closeHandler = () => {
        cleanup();
        if (!chatCompleted && fullReply) {
          this.logger.info('[agentoracle-native-plugin] Connection closed, assuming completion');
          resolve(fullReply);
        } else if (!chatCompleted) {
          reject(new Error('Connection closed before completion'));
        }
      };

      // 超时检查
      const timeoutCheck = setInterval(() => {
        const elapsed = Date.now() - startTime;
        if (elapsed > this.config.timeout * 1000) {
          cleanup();
          reject(new Error(`Task timeout (${this.config.timeout}s)`));
        }

        // 检查是否长时间没有更新
        const timeSinceUpdate = Date.now() - lastUpdateTime;
        if (timeSinceUpdate > this.config.messageTimeout * 2 * 1000) {
          if (chatCompleted || fullReply) {
            this.logger.info('[agentoracle-native-plugin] No updates for a while, assuming completion');
            cleanup();
            resolve(fullReply);
          }
        }
      }, 1000);

      const cleanup = () => {
        clearInterval(timeoutCheck);
        ws.off('message', messageHandler);
        ws.off('error', errorHandler);
        ws.off('close', closeHandler);
      };

      ws.on('message', messageHandler);
      ws.on('error', errorHandler);
      ws.on('close', closeHandler);
    });
  }

  /**
   * 接收单个消息
   * @param ws WebSocket 实例
   * @param timeout 超时时间（毫秒）
   * @returns 解析后的消息对象
   * @private
   */
  private async receiveMessage<T>(ws: WebSocket, timeout: number): Promise<T> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        ws.off('message', messageHandler);
        reject(new Error('Message receive timeout'));
      }, timeout);

      const messageHandler = (data: WebSocket.Data) => {
        clearTimeout(timer);
        ws.off('message', messageHandler);
        try {
          const message = JSON.parse(data.toString());
          resolve(message as T);
        } catch (error) {
          reject(new Error('Failed to parse message'));
        }
      };

      ws.on('message', messageHandler);
    });
  }

  /**
   * 睡眠函数
   * @param ms 毫秒数
   * @private
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
